specs/GetSecurityDescriptorForAppContainerNames/GetSecurityDescriptorForAppContainerNames.md

```diff
@@ -0,0 +1,87 @@
+# Easing and streamlining named object sharing between packaged and Win32
+
+This feature is a flat C API that takes a list of Package Family Names (PFNs) and access masks, and returns a security descriptor
+
+# Background
+
+Named objects (mutexes, events, waitable timers) must be ACLed during creation in order to be shared with packaged applications.
+
+Currently, [our guidance](https://docs.microsoft.com/en-us/windows/uwp/communication/sharing-named-objects) for this scenario is to write a large amount of complex, cumbersome, and largely boilerplate code. ([See here for a taste](https://docs.microsoft.com/en-us/windows/win32/api/securityappcontainer/nf-securityappcontainer-getappcontainernamedobjectpath#examples)).
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_security_descriptor sd;
+    AppContainerAccess access[1] =
+      {{L"Contoso.Test.App_12345678",
+        EVENT_MODIFY_STATE | SYNCHRONIZE}};
+
+    THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
+        1, access, EVENT_MODIFY_STATE | SYNCHRONIZE, &sd);
+
+    SECURITY_ATTRIBUTES sa;
+    sa.nLength = sizeof(sa);
+    sa.lpSecurityDescriptor = sd.get();
+    sa.bInheritHandle = FALSE;
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, &sa);
+    // aka, ::CreateEvent(&sa, FALSE, FALSE, name);
+    return result;
+}
+```
+
+# API Details
+
+```c
+struct AppContainerAccess
+{
+    PCWSTR appContainerName;
+    uint32_t accessMask;
+};
+
+STDAPI GetSecurityDescriptorForAppContainerNames(
+    uint32_t countOfAppContainerNames,
+    _In_reads_(countOfAppContainerNames)
+        const AppContainerAccess* appAccess,
+    uint32_t userAccessMask,
```

Comment:
Not clear how this is used? Each struct has its own access mask.

Reply:
Took me a while to figure this out too. This last one is the access mask for medium IL.

Reply:
DISCUSSION: Sample code on MSDN fetches a PSID and then uses that with this parameter to say "that sid has this access."  The security system effectively makes two passes - one for the low-IL user token, one for the app container identity. It's important to have 

RECOMMEND: Add parameter which is a `PSID principal` and change to `uint32_t principalAccessMask` ... the caller can use one of the WIL helpers (or their own thing) to fetch a `PSID`. Accept `nullptr` to mean "the principal of the current thread/process."

Reply:
Is there any distinction between the principal of the current thread, and the principal of the current process? In other words, can they be different, and if so, which should we use?

Reply:
You should use the one from the current thread, which enables this API to be used correctly on threads that do impersonation (like services and OOP COM and brokers.)